File: quiver2hexo.js
const fs = require('fs');
const shell = require('shelljs');
const rimraf = require('rimraf');
const TurndownService = require('turndown');

const toMdService = TurndownService();

// 源路径
const FROM = '/Users/go_songs/Documents/liuzhiyuan/QuiverData/github.qvnotebook';
// 模板路径
const TO = '/Users/go_songs/Documents/liuzhiyuan/blog/yuan-blog/source';

function initSource(TO) {
  const post = TO + '/_posts';
  const assets = TO + '/assets';
  shell.rm('-rf', post);
  shell.rm('-rf', assets);
  shell.mkdir(post);
  shell.mkdir(assets);
}

function ls(path) {
  const files = fs.readdirSync(path);
  return files.map(file => {
    return path + '/' + file;
  })
}

function cp(filePath, dirPath) {
  if (!hasDir(dirPath)) {
    fs.mkdirSync(dirPath);
  }
  fs.createReadStream(filePath).pipe(fs.createWriteStream(dirPath));
}

function hasDir(path) {
  return fs.existsSync(path);
}

function isDir(path) {
  if (!hasDir(path)) return false;
  return fs.lstatSync(path).isDirectory();
}

function isFile(path) {
  return fs.lstatSync(path).isFile();
}

function readJSON(path) {
  const buffer = fs.readFileSync(path);
  return JSON.parse(buffer.toString('utf8'));
}

function readNote(path) {
  const content = readJSON(path + '/content.json');
  const meta = readJSON(path + '/meta.json');
  const note = {
    uuid: meta.uuid,
    title: meta.title.replace(/:/g, ' ') || '无标题笔记',
    tags: meta.tags,
    date: meta.created_at,
    content: ''
  };

  // 解析 cells
  if (content.cells) {
    content.cells.map(function (cell) {
      if (cell.type === 'text') {
        let content = toMdService.turndown(cell.data) + '\n\n';
        note.content += content.replace(/quiver-image-url/g, '/assets');
      }
      if (cell.type === 'markdown') {
        note.content += cell.data + '\n\n';
      }
      if (cell.type === 'code') {
        note.content += '```' + cell.language + '\n' + cell.data + '\n```\n\n';
      }
    })
  }
  writeMd(note);
}

function writeMd(note) {
  const meta = `---\ntitle: ${note.title}\ndate: 2017-12-15 23:19:13\ntags: ${JSON.stringify(note.tags)}\n---\n\n`;
  fs.writeFileSync(`${TO}/_posts/${note.uuid}.md`, meta + note.content);
}

function writeRes(path) {
  const resPath = path + '/resources';
  if (isDir(resPath)) {
    const res = ls(resPath);
    if (res) {
      res.map(item => {
        shell.cp('-rf', item, TO + '/assets/');
      })
    }
  }
}

function start() {
  initSource(TO);

  ls(FROM).map(function (notePath) {
    if (isDir(notePath)) {
      readNote(notePath);
      writeRes(notePath);
    }
  });
}

start(TO);

File: quiver2hexo.py
import re
import os
import time
import shutil
import json
import html2text
from progress.bar import Bar

# Please specify your quiver notebook path
book_path = './sample.qvnotebook'
# Please specify your Hexo source path
output_path = './source'


post_path = f'{output_path}/_posts'
resource_path = f'{output_path}/images'
h = html2text.HTML2Text()


def reset_output_dir():
    # create output dir if not exist
    if os.path.isdir(output_path):
        shutil.rmtree(output_path)

    os.mkdir(output_path)
    os.mkdir(post_path)
    os.mkdir(resource_path)


def parse_json_file(path):
    if not os.path.isfile(path):
        print(f"ERROR: {path} not exist!")
        exit(0)

    data = {}
    with open(path, 'r') as f:
        data = json.load(f)
        f.close()
    return data


def migrate_book():
    book = parse_json_file(f"{book_path}/meta.json")
    print(f"Migration start {book['name']}...\n")

    note_paths = os.listdir(book_path)
    if len(note_paths) == 0:
        print("WARNING: book is empty!")
        exit(0)

    bar = Bar('Processing: ', max=len(note_paths))

    for note_path in note_paths:
        if os.path.isdir(f"{book_path}/{note_path}"):
            meta = parse_json_file(f"{book_path}/{note_path}/meta.json")
            meta_title = meta['title'] or '无标题笔记'
            meta_tags = meta['tags'] or ''
            meta_date = time.strftime(
                '%Y-%m-%d %H:%M:%S',
                time.localtime(meta['created_at'])
            )
            content = parse_json_file(f"{book_path}/{note_path}/content.json")

            # copy resources
            res_path = f"{book_path}/{note_path}/resources"
            if os.path.isdir(res_path):
                for res_file in os.listdir(res_path):
                    shutil.copy(
                        f"{res_path}/{res_file}",
                        f"{resource_path}/{res_file}"
                    )

            # handle multiple content
            md = ''
            for cell in content['cells']:
                c_type = cell['type']
                c_data = cell['data']
                if c_type == 'markdown':
                    md += c_data
                if c_type == 'text':
                    # replace quiver url to hexo style
                    filter_data = re.sub(
                        r'quiver-image-url/|quiver-file-url/',
                        '/images/',
                        c_data
                    )
                    md += h.handle(filter_data)
                if c_type == 'code':
                    md += f"\n```{cell['language']}\n{c_data}\n```"
                if c_type == 'diagram':
                    md += f"\n```{cell['diagramType']}\n{c_data}\n```"
                md += '\n'

            note_name = note_path.replace('.qvnote', '')
            post_name = f"{post_path}/{note_name}.md"
            with open(post_name, 'w') as f:
                f.truncate()
                f.write(
                    f"""---\ntitle: {meta_title}\ndate: {meta_date}\ntags: {meta_tags}\n---\n\n""")
                f.write(md)
                f.close()

            bar.next()
            time.sleep(0.3)
    bar.finish()
    print("\nMigration success! 👌")


if __name__ == '__main__':
    reset_output_dir()
    migrate_book()
